### PR TITLE
feat (asset) Add summary option for asset query type

### PR DESCRIPTION
**Pull request by OliverCosma**
# Pull Request

## 🤨 Rationale

US: [User Story 2855720](https://dev.azure.com/ni/DevCentral/_workitems/edit/2855720): [UI][Grafana] Create 'Summary' option as part of the 'SystemLink Asset' Grafana plugin

## 👩‍💻 Implementation

Added query to /asset-summary. Modified QueryEditor to not initialize the query with ListAssets every time a different visualisation is selected (the change from useState part).

## 🧪 Testing

Added unit tests for asset summary part.
Manually tested:

grafanaQuery

## ✅ Checklist

- [x] This PR has a title that follows the [commit message format](https://github.com/ni/systemlink-grafana-plugins#commit-message-format).

### CODE REVIEW

**Review by saradei-ni — APPROVED** [review]

File: src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts [hunk 1]
```diff
@@ -17,10 +19,35 @@ export class AssetSummaryDataSource extends AssetDataSourceBase {
...
+        ];
+        return result;
+    }
+
+    async getAssetSummary(): Promise<AssetSummaryQuery> {
+        try {
+            let response = await this.get<AssetSummaryQuery>(this.baseUrl + '/asset-summary');
+            return response;
+        } catch (error) {
+            throw new Error(`An error occurred while getting asset summary: ${error}`);
```
Comment (kkerezsi) [inline src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts:50]: no need to rethrow

File: src/datasources/asset/components/AssetQueryEditor.tsx [hunk 2]
```diff
@@ -56,14 +58,22 @@ export function AssetQueryEditor({ query, onChange, onRunQuery, datasource }: Pr
...
   }, [queryType, handleQueryChange]);
 
   return (
     <div style={{ position: 'relative' }}>
       <InlineField label="Query type" labelWidth={22} tooltip={tooltips.queryType}>
-        <Select options={queryTypeOptions} onChange={handleQueryTypeChange} value={queryType} width={85} />
+        <Select
+          options={queryTypeOptions}
+          onChange={handleQueryTypeChange}
+          value={queryTypeOptions.find(option => option.value === queryType)}
```
Comment (kkerezsi) [inline src/datasources/asset/components/AssetQueryEditor.tsx:74]: Was this change necessary ?

File: src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts [hunk 3]
```diff
@@ -17,10 +19,35 @@ export class AssetSummaryDataSource extends AssetDataSourceBase {
...
-    shouldRunQuery(query: AssetQuery): boolean {
-        throw new Error("Method not implemented.");
+
+    async processMetadataQuery(query: AssetQuery) {
+        const result: DataFrameDTO = { refId: query.refId, fields: [] };
+
+        const assets: AssetSummaryQuery = await this.getAssetSummary();
+
+        result.fields = [
+            { name: 'Total', values: [assets.total] },
```
Comment (kkerezsi) [inline src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts:36]: Since we are reusing these values in the tests, we might want to create a constants for this

File: src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts [hunk 4]
```diff
@@ -17,10 +19,35 @@ export class AssetSummaryDataSource extends AssetDataSourceBase {
...
+
+    shouldRunQuery(_: AssetQuery): boolean {
+        return true;
     }
-    shouldRunQuery(query: AssetQuery): boolean {
-        throw new Error("Method not implemented.");
+
+    async processMetadataQuery(query: AssetQuery) {
+        const result: DataFrameDTO = { refId: query.refId, fields: [] };
+
```
Comment (kkerezsi) [inline src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts:32]: we can remove this empty line

**Review by kkerezsi — APPROVED** [review]

File: src/datasources/asset/components/AssetQueryEditor.tsx (hunk 2 above)
Comment (OliverCosma) [reply src/datasources/asset/components/AssetQueryEditor.tsx:74]: Yes, due to the `setState` change which now points to a string and not AssetQueryType

File: src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts (hunk 3 above)
Comment (OliverCosma) [reply src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts:36]: I'm not sure if we should change it now since we have this just here once and once in the test. If we'll add more, we might consider adding a constant

File: src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts (hunk 1 above)
Comment (saradei-ni) [reply src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts:50]: Isn't this how we handled the main requests in the other data sources as well? Should we change that approach?

Comment (OliverCosma) [reply src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts]: @kkerezsi  I modified the request for my PR, but just wanted to let you know that we have this approach everywhere. Should we create a tech debt to address this?

File: src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts (hunk 3 above)
Comment (OliverCosma) [reply src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts]: Added constant

File: src/datasources/asset/components/AssetQueryEditor.tsx [hunk 5]
```diff
@@ -31,8 +31,11 @@ export function AssetQueryEditor({ query, onChange, onRunQuery, datasource }: Pr
   }, [onChange, onRunQuery]);
 
   const handleQueryTypeChange = useCallback((item: SelectableValue<AssetQueryType>): void => {
+    if(query.queryType === item.value){
```
Comment (kkerezsi) [inline src/datasources/asset/components/AssetQueryEditor.tsx:34]: @OliverCosma make sure you remove this

Comment (kkerezsi) [reply src/datasources/asset/components/AssetQueryEditor.tsx]: Sorted this one out for you

File: src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts (hunk 1 above)
Comment (kkerezsi) [reply src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts]: Let's leave it as is for now and we'll treat it as part of a generic code since it will anyways not display the right thing. 

The error message will be 'An error occurred while getting asset summary: [Object object]'

I've created a task for this

File: src/datasources/asset/components/AssetQueryEditor.tsx [hunk 6]
```diff
@@ -66,7 +66,12 @@ export function AssetQueryEditor({ query, onChange, onRunQuery, datasource }: Pr
   return (
     <div style={{ position: 'relative' }}>
       <InlineField label="Query type" labelWidth={22} tooltip={tooltips.queryType}>
-        <Select options={filterOptions} onChange={handleQueryTypeChange} value={queryType} width={85} />
+        <Select
```
Comment (mure) [inline src/datasources/asset/components/AssetQueryEditor.tsx:69]: I guess I missed this in an earlier change, but this "Query type" field doesn't match the UX we have in other data sources. Is that intentional? We've used the `RadioButtonGroup` in the past:
https://github.com/ni/systemlink-grafana-plugins/blob/8a028a82a2688d105eb2e165fcf83a558e2069ba/src/datasources/tag/components/TagQueryEditor.tsx#L38

File: src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts [hunk 7]
```diff
@@ -17,10 +19,29 @@ export class AssetSummaryDataSource extends AssetDataSourceBase {
...
+        return this.processMetadataQuery(query as AssetQuery);
+    }
+
+    shouldRunQuery(_: AssetQuery): boolean {
+        return true;
     }
-    shouldRunQuery(query: AssetQuery): boolean {
-        throw new Error("Method not implemented.");
+
+    async processMetadataQuery(query: AssetQuery) {
```
Comment (mure) [inline src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts:30]: Shouldn't this be `processSummaryQuery()`?

File: src/datasources/asset/types/AssetSummaryQuery.types.ts [hunk 8]
```diff
@@ -1,4 +1,9 @@
 import { DataQuery } from '@grafana/schema'
 
 export interface AssetSummaryQuery extends DataQuery {
-}
+    total: number;
+    active: number;
+    notActive: number;
+    approachingRecommendedDueDate: number;
+    pastRecommendedDueDate: number;
+  }
```
Comment (mure) [inline src/datasources/asset/types/AssetSummaryQuery.types.ts:9]: This doesn't make sense—the `DataQuery` type is meant to represent the query a user builds for a data source. This type is more like an `AssetSummaryResponse`.

File: src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts [hunk 9]
```diff
@@ -17,10 +19,29 @@ export class AssetSummaryDataSource extends AssetDataSourceBase {
...
+        const result: DataFrameDTO = { refId: query.refId, fields: [] };
```
Comment (mure) [inline src/datasources/asset/components/editors/asset-summary/AssetSummaryDataSource.ts:31]: nit: I don't see the point of defining this variable and the setting the fields in two steps. Just return the whole thing at the end:

```ts
return {
  refId: query.refId,
  fields: [
    { name: assetSummaryFields.TOTAL, values: [assets.total] },
    ...
  ]
}
```

**Review by mure — COMMENTED** [review]
The models we're using to save Asset queries in the dashboard JSON is wrong. We need to be saving the type of the query like we do here: https://github.com/ni/systemlink-grafana-plugins/blob/6a8001277d86bfe758fc18eeafffa11b5774fbc4/src/datasources/tag/types.ts#L9

For example, there's no way for the data source to know that this is a calibration forecast query:

https://github.com/ni/systemlink-grafana-plugins/blob/6a8001277d86bfe758fc18eeafffa11b5774fbc4/src/datasources/asset/types/CalibrationForecastQuery.types.ts#L4

File: src/datasources/asset/components/AssetQueryEditor.tsx (hunk 6 above)
Comment (kkerezsi) [reply src/datasources/asset/components/AssetQueryEditor.tsx:69]: @mure this was already discussed with @TJ-G

**Comment by kkerezsi** [comment]
@mure Type changes are going to be part of a separate PR.

**Review by mure — APPROVED** [review]
